File: eulerglot/elixir/lib/003_prime_factors.ex
defmodule PrimeFactors do

  @moduledoc """
  The prime factors of 13195 are 5, 7, 13 and 29.
  What is the largest prime factor of the number 600851475143 ?
  """

  def largest(number) when number < 1, do: 0
  def largest(number) do
    div_by(number, 2)
    |> factor_prime(3)
  end

  defp factor_prime(1, _), do: 2
  defp factor_prime(number, prime) do
    {number, largest} = factor_prime(number, prime, 0)
    if number > 2, do: number, else: largest
  end

  defp factor_prime(number, prime, largest) do
    if prime > :math.sqrt(number) do
      {number, largest}
    else
      number = div_by(number, prime)
      factor_prime(number, prime + 2, prime)
    end
  end

  defp div_by(number, multiple) when rem(number, multiple) != 0, do: number
  defp div_by(number, multiple) do
    (number / multiple) |> round |> div_by(multiple)
  end

end

File: eulerglot/clojure/src/clojure_euler/003_prime_factors.clj
(ns clojure-euler.003_prime_factors)

(comment "The prime factors of 13195 are 5, 7, 13 and 29.
  What is the largest prime factor of the number 600851475143 ?")

(defn factor-by
  [number multiple]
  (if (not= 0 (rem number multiple))
    number
    (recur (/ number multiple) multiple)))

(defn factor-by-prime
  [number prime largest]
  (if (> prime (Math/sqrt number))
    [number largest]
    (recur (factor-by number prime) (+ prime 2) prime)))

(defn largest
  [ceil]
  (let [number (factor-by ceil 2)]
  (let [result (factor-by-prime number 3 0)]
  (let [number (nth result 0) largest (nth result 1)]
  (if (> number 2) number largest)))))
